src/predictor/gpu_predictor.cu

```diff
@@ -324,22 +336,31 @@ class GPUPredictor : public xgboost::Predictor {
                 h_nodes.begin() + h_tree_segments[tree_idx - tree_begin]);
     }
 
-    size_t i_batch = 0;
     size_t batch_offset = 0;
     for (auto &batch : dmat->GetRowBatches()) {
-      CHECK(i_batch == 0 || devices_.Size() == 1) << "External memory not supported for multi-GPU";
-      // out_preds have been sharded and resized in InitOutPredictions()
-      batch.offset.Shard(GPUDistribution::Overlap(devices_, 1));
+      GPUSet devices = GPUSet::All(param_.gpu_id, param_.n_gpus, batch.Size());
+      ConfigureShards(devices);
+      std::vector<size_t> out_preds_offsets;
+      PredictionDeviceOffsets(out_preds->Size(), batch_offset, batch.Size(),
+                              model.param.num_output_group, &out_preds_offsets);
+      out_preds->Reshard(GPUDistribution::Explicit(devices_, out_preds_offsets));
+
+      batch.offset.Reshard(GPUDistribution::Overlap(devices_, 1));
       std::vector<size_t> device_offsets;
       DeviceOffsets(batch.offset, &device_offsets);
       batch.data.Reshard(GPUDistribution::Explicit(devices_, device_offsets));
+
       dh::ExecuteIndexShards(&shards_, [&](int idx, DeviceShard& shard) {
-        shard.PredictInternal(batch, dmat->Info(), out_preds, batch_offset, model,
+        shard.PredictInternal(batch, dmat->Info(), out_preds, model,
```

Comment:
This actually copies the trees to the GPU for every SparsePage. Don't we want to copy them once instead?

Reply:
Good point. Added a TODO, will send a followup PR.